pennylane/io/qualtran_io.py

```diff
@@ -0,0 +1,198 @@
+# Copyright 2018-2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This submodule contains the adapter class for Qualtran-PennyLane interoperability.
+"""
+# pylint:disable=
+from qualtran import (
+    Bloq,
+    CompositeBloq,
+    Soquet,
+    LeftDangle,
+    Side,
+    DecomposeNotImplementedError,
+    DecomposeTypeError,
+)
+
+import numpy as np
+import pennylane as qml
+
+from pennylane.operation import Operation
+from pennylane.wires import WiresLike
+
+
+def get_bloq_registers_info(bloq):
+    """Returns a `qml.registers` object associated with all named and unnamed registers and wires
+    in the bloq.
+
+    Args:
+        bloq: the bloq to get the registers info of
+
+    Returns:
+        dict: A dictionary that has all the named and un-named registers with default wire
+        ordering.
+
+    **Example**
+
+    Given a qualtran bloq:
+
+    from qualtran.bloqs.basic_gates import Swap
+
+    >>> qml.get_bloq_registers_info(Swap(3))
+    {'x': Wires([0, 1, 2]), 'y': Wires([3, 4, 5])}
+    """
+
+    cbloq = bloq.decompose_bloq() if not isinstance(bloq, CompositeBloq) else bloq
+
+    temp_register_dict = {}
+    for reg in cbloq.signature.rights():
+        temp_register_dict[reg.name] = reg.bitsize
+
+    return qml.registers(temp_register_dict)
+
+
+def _get_named_registers(registers):
+    """Returns a `qml.registers` object associated with the named registers in the bloq"""
+
+    temp_register_dict = {}
+    for reg in registers:
+        temp_register_dict[reg.name] = reg.bitsize
+
+    return qml.registers(temp_register_dict)
+
+
+class FromBloq(Operation):
+    r"""
+    A shim for using bloqs as a PennyLane operation.
+
+    Args:
+        bloq: the bloq to wrap
+        wires: the wires to act on
+
+    **Example**
+
+    Given a qualtran bloq:
+
+    from qualtran.bloqs.basic_gates import CNOT
+
+    >>> qualtran_toffoli = qml.FromBloq(CNOT(), [0, 1])
+    >>> qualtran_toffoli.matrix()
+    array([[1.+0.j, 0.+0.j, 0.+0.j, 0.+0.j],
+       [0.+0.j, 1.+0.j, 0.+0.j, 0.+0.j],
+       [0.+0.j, 0.+0.j, 0.+0.j, 1.+0.j],
+       [0.+0.j, 0.+0.j, 1.+0.j, 0.+0.j]])
+
+    A simple example showcasing how to use `qml.FromBloq` inside a device:
+
+    .. code-block::
+
+        dev = qml.device("default.qubit")
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.FromBloq(XGate(), [0])
+            return qml.expval(qml.Z(wires=[0]))
+
+    >>> circuit()
+    -1.0
+    """
+
```

Comment:
we might need a class variable like `_has_qualtran = has_qualtran` that can be used in the `__init__` to do like -
```suggestion
	if self._has_qualtran:
        raise ImportError(
            "This feature requires qualtran, a quantum algorithms translator. "
            "It can be installed with:\n\npip install qualtran"
        )
```

Reply:
This import error wouldn't run due to the try-except block at the top of the file. AKA no code coverage so I removed it.